libparsec/crates/client_connection/tests/proxy/mod.rs

```diff
@@ -0,0 +1,169 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+// `allow-unwrap-in-test` don't behave as expected, see:
+// https://github.com/rust-lang/rust-clippy/issues/11119
+#![allow(clippy::unwrap_used)]
+
+use std::io;
+
+use libparsec_platform_http_proxy::ProxyConfig;
+use libparsec_types::BackendAddr;
+use rand::Rng;
+use tokio::{
+    net::{TcpListener, TcpStream},
+    sync::{mpsc, oneshot},
+};
+
+pub struct ProxyHandle {
+    join_handle: tokio::task::JoinHandle<io::Result<()>>,
+    port: u16,
+    notify_disconnect: mpsc::Sender<Event>,
+}
+
+impl ProxyHandle {
+    pub fn get_proxy(&self) -> anyhow::Result<ProxyConfig> {
+        ProxyConfig::default()
+            .with_http_proxy(format!("http://localhost:{}", self.port))?
+            .with_https_proxy(format!("http://localhost:{}", self.port))
+    }
+
+    pub async fn disconnect(&self) {
+        log::trace!("Notify proxy server to close the connections ...");
+        self.notify_disconnect
+            .send(Event::Disconnect)
+            .await
+            .unwrap()
+    }
+
+    pub async fn close(self) -> io::Result<()> {
+        self.notify_disconnect.send(Event::Close).await.unwrap();
+        self.join_handle.await.unwrap()
+    }
+}
+
+enum Event {
+    Close,
+    Disconnect,
+}
+
+pub async fn spawn(backend_addr: BackendAddr) -> io::Result<ProxyHandle> {
+    let port = rand::thread_rng().gen_range(10_000..u16::MAX);
```

Comment:
Specify a port at 0 in `TcpListener::bind` and get back to allocated port with `TcpListener::local_addr` instead

Reply:
done